### PR TITLE
Add the command line tool kwokctl

### PR DESCRIPTION
**Pull request by wzshiming**
# Feature

``` console
Kwokctl is a Kwok cluster management tool

Usage:
  kwokctl [command] [flags]
  kwokctl [command]

Available Commands:
  completion  Generate the autocompletion script for the specified shell
  create      Creates one of [cluster]
  delete      Deletes one of [cluster]
  get         Gets one of [artifacts, clusters, kubeconfig]
  help        Help about any command
  kubectl     kubectl in cluster
  logs        Logs one of [etcd, kube-apiserver, kube-controller-manager, kube-scheduler, kwok-controller, prometheus]

Flags:
  -h, --help          help for kwokctl
      --name string   cluster name (default "kwok")

```

## Runtime and OS support

|         | binary | docker | nerdctl | kind |
| ------- | ------ | ------ | ------- | ---- |
| Linux   | ✔️      | ✔️      | ✔️       | ✔️    |
| Darwin  | ➕      | ✔️      | ✖️       | ✔️    |
| Windows | ➕      | ✔️      | ✖️       | ✔️    |

✔️: Support
✖️: Not support
➕: You need to use your own build of the kube binary

Not tested on Windows yet.

## Default Runtime

Linux: binary
Darwin/Windows: docker

## All-in-one image

``` bash
./images/cluster/build.sh --tag cluster:v0.1.0-kube-1.24 --kube-version 1.24.2
docker run --rm -it -p 8080:8080 cluster:v0.1.0-kube-1.24
```

### CODE REVIEW

**Comment by wzshiming** [comment]
/hold
Waiting for review

**Comment by wzshiming** [comment]
The `test/kwokctl/kwokctl.test.sh` for testing the features in this commit

File: pkg/kwokctl/cmd/kwokctl/get/images/images.go
Comment (yibozhuang) [inline pkg/kwokctl/cmd/kwokctl/get/images/images.go:40]: same here, get the list of supported runtimes from the runtime pkg

File: pkg/kwokctl/cmd/kwokctl/get/kubeconfig/kubeconfig.go
Comment (yibozhuang) [inline pkg/kwokctl/cmd/kwokctl/get/kubeconfig/kubeconfig.go:65]: nit: this can be simplified to:
```go
return err
```

same for the other places

File: pkg/kwokctl/pki/openssl.cnf
Comment (yibozhuang) [inline pkg/kwokctl/pki/openssl.cnf:14]: nit: new line

File: pkg/kwokctl/pki/admin.crt
Comment (yibozhuang) [inline pkg/kwokctl/pki/admin.crt:1]: do we need to check in `admin.crt`, `admin.key`, and `ca.crt` if they can be generated by `DumpPki()` in `pki.go`?

File: pkg/kwokctl/runtime/binary/cluster.go
Comment (yibozhuang) [inline pkg/kwokctl/runtime/binary/cluster.go:424]: for later: maybe we can think about doing the `ForkExecKill` of each component in parallel rather than serial (same for bring up)

File: pkg/kwokctl/runtime/binary/prometheus.yaml.tpl
Comment (yibozhuang) [inline pkg/kwokctl/runtime/binary/prometheus.yaml.tpl:111]: new line.

File: pkg/kwokctl/runtime/compose/prometheus.yaml.tpl
Comment (yibozhuang) [inline pkg/kwokctl/runtime/compose/prometheus.yaml.tpl:111]: new line.

File: pkg/kwokctl/utils/cmd.go
Comment (yibozhuang) [inline pkg/kwokctl/utils/cmd.go:164]: need to handle error here.

File: pkg/kwokctl/utils/cmd.go
Comment (yibozhuang) [inline pkg/kwokctl/utils/cmd.go:37]: should you also call `os.FindProcess()` to ensure that it is already running?

otherwise there might be a case where if the kill failed and we do not remove the pid and that would cause inconsistencies.

File: pkg/kwokctl/utils/cmd.go
Comment (yibozhuang) [inline pkg/kwokctl/utils/cmd.go:113]: so if the kill failed here, we would not remove the pid path, do we want to have this file remaining?

File: pkg/kwokctl/utils/download.go
Comment (yibozhuang) [inline pkg/kwokctl/utils/download.go:59]: need to handle error here.

File: pkg/kwokctl/utils/cmd.go
Comment (wzshiming) [reply pkg/kwokctl/utils/cmd.go:113]: Yes
In `killProcess`, kill process success and process not exist should return nil, 
if it returns an error I think we should keep the pid path for later.

File: pkg/kwokctl/pki/admin.crt
Comment (wzshiming) [reply pkg/kwokctl/pki/admin.crt:1]: Is the check format legal?

Comment (yibozhuang) [reply pkg/kwokctl/pki/admin.crt]: I mean like, do these files need to be part of the PR? Since they are generated.

File: pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go
Comment (Huang-Wei) [inline pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go:180]: 💭 to brainstorm: should we cleanup the leftover once an error happens? How do other tools (like kind) deal with this situation?

File: pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go
Comment (Huang-Wei) [inline pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go:109]: Is `tar` intentional? I think we should accept binary path as the input, either to replace `tar` or coexist.

And an 💭 is to lookup the default path of etcd/prometheus and use it by default, if not specified.

Comment (Huang-Wei) [reply pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go:109]: for Mac M1, there is no prebuild etcd image/binary lol...

https://github.com/etcd-io/etcd/releases/tag/v3.5.4

- [etcd-v3.5.4-darwin-amd64.zip](https://github.com/etcd-io/etcd/releases/download/v3.5.4/etcd-v3.5.4-darwin-amd64.zip)
- [etcd-v3.5.4-linux-amd64.tar.gz](https://github.com/etcd-io/etcd/releases/download/v3.5.4/etcd-v3.5.4-linux-amd64.tar.gz)
- [etcd-v3.5.4-linux-arm64.tar.gz](https://github.com/etcd-io/etcd/releases/download/v3.5.4/etcd-v3.5.4-linux-arm64.tar.gz)
- [etcd-v3.5.4-linux-ppc64le.tar.gz](https://github.com/etcd-io/etcd/releases/download/v3.5.4/etcd-v3.5.4-linux-ppc64le.tar.gz)
- [etcd-v3.5.4-linux-s390x.tar.gz](https://github.com/etcd-io/etcd/releases/download/v3.5.4/etcd-v3.5.4-linux-s390x.tar.gz)
- [etcd-v3.5.4-windows-amd64.zip](https://github.com/etcd-io/etcd/releases/download/v3.5.4/etcd-v3.5.4-windows-amd64.zip)

File: pkg/kwokctl/utils/cmd_unix.go
Comment (Huang-Wei) [inline pkg/kwokctl/utils/cmd_unix.go:1]: should we name this file to replace `cmd.go`? Or, two files named "cmd_other.go" and "cmd_windows.go".

File: pkg/kwokctl/utils/filepath_unix.go
Comment (Huang-Wei) [inline pkg/kwokctl/utils/filepath_unix.go:1]: given it's non-windows, I'd suggest to rename it as `filepath_other.go`.

**Review by Huang-Wei — COMMENTED** [review]
Thanks for this significant PR.

First found of review, but overall it's superb! Will test it locally after the "etcd binary path" parameter gets introduced (or "binary lookup logic" can be introduced).

File: pkg/kwokctl/runtime/registry.go
Comment (wzshiming) [reply pkg/kwokctl/runtime/registry.go:47]: I think it's `docker compose`, which is the acronym I used for my first runtime implementation, and it's been copied ever since. I'll update it.

File: pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go
Comment (wzshiming) [reply pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go:154]: Failed to download binary or pull an image, or ran out of disk space.

File: pkg/kwokctl/runtime/kind/cluster.go
Comment (wzshiming) [reply pkg/kwokctl/runtime/kind/cluster.go:34]: Yes
`*runtime.Cluster` is a common implementation of the interface by different Runtimes, such as saving and loading the current cluster configuration, checking if the cluster is ready.
I think it's easy to misunderstand, I'll move it to `runtime/common`.

File: pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go
Comment (wzshiming) [reply pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go:148]: The only reason to get here is that the cluster already exists, but is not ready for various reasons
`Install` is for an interruption halfway through the installation and then needs to continue the installation.
`Down` is for an interruption halfway through startup and then shutting down the already started one.

File: pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go
Comment (wzshiming) [reply pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go]: Is tar intentional? I think we should accept binary path as the input, either to replace tar or coexist.
Etcd and Prometheus releases do not provide a binary

        And an 💭 is to lookup the default path of etcd/prometheus and use it by default, if not specified.

That makes sense, I'll see how to do it right.

File: pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go
Comment (wzshiming) [reply pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go]: Kind no cleanup after a failed cluster creation, to save the scene for troubleshooting.

File: pkg/kwokctl/runtime/kind/cluster.go
Comment (Huang-Wei) [reply pkg/kwokctl/runtime/kind/cluster.go:34]: ok, I understood it now. `Cluster` has different implementations on different "runtime".

File: pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go
Comment (Huang-Wei) [reply pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go]: Upon these errors, shouldn't we abort and quit? Otherwise, (re-)install would fail again?

File: pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go
Comment (Huang-Wei) [reply pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go]: > The only reason to get here

IIUC, when rt.Config(), rt.Ready(ctx) and rt.Install(ctx, *conf) all runs well and return nil err, the logic Down() is hit. so literally, it sounds to me it's a happy path, and why we call `Down()` in a happy path.

May there is something missing in the semantics of `err`. If so, I'd suggest to refactor the `err` as semantical `status`, so the readers can easily understand and distinguish the cases like "cluster already configured", etc.

File: pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go
Comment (Huang-Wei) [reply pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go]: We can do it in a follow-up.

Pretty nice, I can install it on my M1 🎉 :

```
⇒  kwokctl create cluster --runtime binary --kube-scheduler-binary /usr/local/bin/kube-scheduler@1.24.3 --kube-controller-manager-binary /usr/local/bin/kube-controller-manager@1.24.3  --kube-apiserver-binary /usr/local/bin/kube-apiserver@1.24.3 --kwok-controller-binary /usr/local/bin/kwok --etcd-binary /opt/homebrew/bin/etcd
Creating cluster "kwok-kwok"
Starting cluster "kwok-kwok"
Cluster "kwok-kwok" is ready
You can now use your cluster with:

    kubectl config use-context kwok-kwok

Thanks for using kwok!
```

File: pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go
Comment (Huang-Wei) [reply pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go]: My $0.02 is that it makes more sense to default to "cleanup upon faulty install". Take my UX for example, yesterday I installed it successfully, and it has a "cluster leftover". And today when I retry (with another sets of parameters), it seems to pickup the old config and failed anyway. It takes me some time to figure out the "leftover config was read" and I have to manually delete it.

cc @yibozhuang for thoughts.

File: pkg/kwokctl/cmd/get/get.go
Comment (Huang-Wei) [inline pkg/kwokctl/cmd/get/get.go:42]: seems `kwokctl get artifacts --name kwok` returns the default images (instead of binaries) that the cluster is using:

```
⇒  kwokctl get artifacts --name kwok
https://dl.k8s.io/release/v1.24.2/bin/darwin/arm64/kubectl
registry.k8s.io/etcd:3.5.4-0
registry.k8s.io/kube-apiserver:v1.24.2
registry.k8s.io/kube-controller-manager:v1.24.2
registry.k8s.io/kube-scheduler:v1.24.2
registry.k8s.io/kwok/kwok:v0.1.0
docker.io/prom/prometheus:v2.35.0
```

The command I used to spin up the cluster was:

```
kwokctl create cluster --runtime binary --kube-scheduler-binary /usr/local/bin/kube-scheduler@1.24.3 --kube-controller-manager-binary /usr/local/bin/kube-controller-manager@1.24.3  --kube-apiserver-binary /usr/local/bin/kube-apiserver@1.24.3 --kwok-controller-binary /usr/local/bin/kwok --etcd-binary /opt/homebrew/bin/etcd
```

File: pkg/kwokctl/cmd/logs/logs.go
Comment (Huang-Wei) [inline pkg/kwokctl/cmd/logs/logs.go:50]:
```suggestion
	cmd.Flags().BoolVarP(&flags.Follow, "follow", "f", false, "Specify if the logs should be streamed.")
```

File: pkg/kwokctl/cmd/delete/delete.go
Comment (Huang-Wei) [inline pkg/kwokctl/cmd/delete/delete.go:29]: FYI: logs when deleting a running cluster using binary as the runtime:

```
⇒  kwokctl delete cluster
Stopping cluster "kwok-kwok"
failed to kill kwok: wait: no child processes
failed to kill kube-scheduler: wait: no child processes
failed to kill kube-controller-manager: wait: no child processes
failed to kill kube-apiserver: wait: no child processes
failed to kill etcd: wait: no child processes
Deleting cluster "kwok-kwok"
Cluster "kwok-kwok" deleted
```

Are the "failed to kill" strings as expected?

**Comment by wzshiming** [comment]
Waiting merged of kubernetes-sigs/kwok#11 and rebase

File: pkg/kwokctl/pki/admin.crt
Comment (yibozhuang) [reply pkg/kwokctl/pki/admin.crt]: @wzshiming just wanted to clarify if the cert files need to be part of the PR if we can generate them?

File: pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go
Comment (yibozhuang) [reply pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go]: Personally I would prefer to have the clean up done on a failed install but keep the log around for users to check.
Additionally, having a switch/flag to allow the user to run the install again without clean up for further debug.

We should be as idempotent as possible meaning that a previously failed case should ideally not affect future retries.

File: pkg/kwokctl/pki/admin.crt
Comment (wzshiming) [reply pkg/kwokctl/pki/admin.crt]: Got it, I feel that adding an extra dependency on OpenSSL is not good, we can add a Go implementation later.

File: pkg/kwokctl/runtime/binary/promehteus.go
Comment (yibozhuang) [inline pkg/kwokctl/runtime/binary/promehteus.go:1]: typo for this filename: `prometheus.go`

File: pkg/kwokctl/runtime/list.go
Comment (yibozhuang) [inline pkg/kwokctl/runtime/list.go:32]:
```suggestion
	var ret []string
```

File: pkg/kwokctl/utils/cmd.go
Comment (yibozhuang) [inline pkg/kwokctl/utils/cmd.go:155]: remove `ctx` since its not being used

File: pkg/kwokctl/utils/cmd_other.go
Comment (yibozhuang) [inline pkg/kwokctl/utils/cmd_other.go:32]: is there a reason we set the sysproc attribute in `startProcess` here where as in `cmd_windows.go` the sysproc attribute for hide window is done in `command`? could we make it consistent?

File: pkg/kwokctl/utils/filepath_windows.go
Comment (yibozhuang) [inline pkg/kwokctl/utils/filepath_windows.go:27]: I'm confused on this: we want to modify the path because these are paths used within Docker/KinD running on windows correct?

What about user specifying `binary` when running on windows?

File: Makefile
Comment (Huang-Wei) [reply Makefile:43]: nit: I think `.DEFAULT_GOAL := all` is the default setting already.

File: pkg/kwokctl/cmd/get/get.go
Comment (wzshiming) [reply pkg/kwokctl/cmd/get/get.go:42]: Fixed

File: pkg/kwokctl/utils/cmd_other.go
Comment (wzshiming) [reply pkg/kwokctl/utils/cmd_other.go:32]: This is the attribute I got after independent testing on both sides, the definition of the sysproc attribute of the two systems is different

File: pkg/kwokctl/utils/filepath_windows.go
Comment (wzshiming) [reply pkg/kwokctl/utils/filepath_windows.go:27]: > I'm confused on this: we want to modify the path because these are paths used within Docker/KinD running on windows correct?

Yes

> What about user specifying binary when running on windows?

Both `\` and `/` work on Windows.

File: pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go
Comment (wzshiming) [reply pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go]: I added logic that will clean up upon faulty install.

File: pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go
Comment (wzshiming) [reply pkg/kwokctl/cmd/kwokctl/create/cluster/cluster.go]: Updated

**Comment by wzshiming** [comment]
Updated for all comments
entrypoint of e2e-test will be modified after this merge, in #11

/unhold

**Review by yibozhuang — APPROVED** [review]

**Comment by Huang-Wei** [comment]
/lgtm

We can fix nits in follow-ups. :shipit:

**Pull request merged**
